### DISCUSSION TITLE
Enhancing Access to CTE Data in SQLAlchemy Models

### DISCUSSION THREAD

**Discussion by FeehGb** [sqlalchemy/sqlalchemy#11384] [Usage Questions]
**Problem Description:**

Currently, I have a model called PriceResearchListModel. However, I'm facing difficulties when trying to add Common Table Expressions (CTE) to it. My goal is to access the data from CTEs when I perform a query using `select(PriceResearchListModel)`. Additionally, I would like to implement data pagination.

The current approach works, but I feel it's not the most appropriate way to handle this. Below is the relevant code snippet:

```python
def _build_select_statement(self, cte_counter: tuple[CTE, ...]) -> Select:
    """
    Builds the select statement for price research lists

    :return: Select statement
    """
    cte_counter_cluster, cte_counter_competitor, cte_counter_family, cte_counter_replicated_family = cte_counter
    self.LPPM.qtd_familia = column_property(cte_counter_family.c.qtd_familia)
    self.LPPM.qtd_familia_inativa = column_property(cte_counter_family.c.qtd_familia_inativa)
    self.LPPM.qtd_familia_nao_encontrada = column_property(
        cte_counter_family.c.qtd_familia_nao_encontrada
    )

    self.LPPM.qtd_concorrente = column_property(cte_counter_competitor.c.qtd_concorrente)
    self.LPPM.qtd_cluster = column_property(cte_counter_cluster.c.qtd_cluster)
    self.LPPM.qtd_familia_replicada = column_property(
        coalesce(cte_counter_replicated_family.c.qtd_familia_replicada, 0)
    )
    self.LPPM.fl_todas_familias_replicada = column_property(
        self.LPPM.qtd_familia_replicada == self.LPPM.qtd_familia
    )

    return select(self.LPPM)
```

**Question:**

Is there an alternative approach to solving this problem? I would appreciate suggestions and guidance on how to improve this implementation.

I have already attempted to follow the documentation at [Subqueries and CTEs](https://docs.sqlalchemy.org/en/20/tutorial/data_select.html#subqueries-and-ctes), but unfortunately, I haven't been successful in implementing the desired functionality.

**Comment by zzzeek** [comment]
those are column_propertys, those are a class-level part of the mapping and would be something you'd only set up once at module import time.     you wouldn't build an ad-hoc `select()` like that at all.

"My goal is to access the data from CTEs when I perform a query using select(PriceResearchListModel)" OK, you can either map PriceResearchListModel to the CTE directly, or you can use `aliased()`, see examples at https://docs.sqlalchemy.org/en/20/tutorial/data_select.html#orm-entity-subqueries-ctes

**Reply by FeehGb** [reply]
Thanks for the guidance. I did try to follow the tutorial, but even after implementing the examples provided, it seems like SQLAlchemy isn't returning all the necessary data. Not only do I require pagination alongside the query results, but it also appears that complete data retrieval isn't happening.

I try mapping PriceResearchListModel directly to the CTE or using aliased() as suggested in the documentation.

**Reply by FeehGb** [reply]
The main problem is the return is
(Model, 2,4 ,5,6) -> where the numbers is the result of cte, but how can I extract this values when i use pydantic ?

**Reply by CaselIT** [reply]
It's really not clear what you are trying to do. What is `LPPM`? what is `cte_counter`?
Can you provide a more complete example?
